lectures_annotated/Week1Lecture2.agda

```diff
@@ -0,0 +1,119 @@
+
+To solve this, we need a list that codifies the length of the list as part of it: morally, we want something that tells us how many conses ,- there are in the list (remember that a 1-element list ["hello"] is actually "hello" ,- []). For this, we need to start by encoding our natural numbers. We can say that every natural number is either zero, or a successor of another natural number.
+-}
+data Nat : Set where
+    zero : Nat
+    suc  : Nat -> Nat
+
+variable
+    m n p : Nat
+
+-- And now we can define our length-specified list.
+data ListN (X : Set) : Nat -> Set where
+    []   : ListN X zero
+    _,-_ : (x : X) (xs : ListN X n) -> ListN X (suc n)
+{-
+Breaking this down, we still define the type of elements in the list (X : Set). But now, instead of defining a type for our data structure, we're generating a *family of types*: one for each natural number. In essence, our types are "length-zero list", "length-one list", "length-two list" and so on.
+We then have our two cases. The empty list [] has length zero. The non-empty list contains an element of type X (the head), a list xs of length n (the tail), and a length of (suc n), ie n + 1.
+-}
+
+
+-- This gives us another safe version of head.
+headN : ListN X (suc n) -> X
+headN (x ,- xs) = x
+{-
+Originally, we were trying to promise "if you give me any list, I'll give you a value", but that's actually too much to promise. We now have a way to express our actual promise, which is "if you give me a non-empty list, I'll give you a value". We can do this by specifying that the length value of our ListN just has to be a successor of something - ie, that our list has at least one element.
+-}
+
+
+-- We also have a safe zap!
+zapN : {T U : Set} -> ListN (T -> U) n -> ListN T n -> ListN U n
+zapN [] []               = []
+zapN (f ,- fs) (t ,- ts) = (f t) ,- zapN fs ts
+{-
+By specifying that the lists that we're zapping *must* be the same length, we can discard our annoying "raggedy" list cases. We will also know the length of our output list - the same as the inputs. This is one example of how Agda's strict typing can end up giving us more information about the data we're working with.
+-}
+
+
+-- We can define append for ListN as well. We will need an understanding of *addition* for natural numbers for this.
+_+N_ : Nat -> Nat -> Nat
+zero  +N m = m  
+suc n +N m = suc (n +N m)
+
+appendN : ListN X n -> ListN X m -> ListN X (n +N m)
+appendN [] ys = ys  
+appendN (x ,- xs) ys = x ,- appendN xs ys
+{-
+Note that here, because we've defined our addition function as being "left to right", so to speak, our definition of the length of our output list has to give the list lengths in "left to right" order - if we give our output's type as "List X (m +N n)" Agda will get mad, because it doesn't know that n+m == m+n (since our addition function lacks reflexivity).
```

Comment:
```suggestion
Note that here, because we've defined our addition function as being "left to right", so to speak, our definition of the length of our output list has to give the list lengths in "left to right" order - if we give our output's type as "List X (m +N n)" Agda will get mad, because it doesn't know that n+m == m+n (since the definition of our addition function is not symmetric: it works by taking the first argument apart).
```